### PR TITLE
topmodel: use multipart uploads for S3FileSystem

### PR DESCRIPTION
**Pull request by mduvall**
r? @alyssafrazee 

### What's in here

This PR implements multipart uploads to avoid the `[Errno 104]` errors from `boto`. It doesn't do anything particularly exciting, I'll point out a few weird things in the code.

This ran successfully on the `trainingbox`s and I was able to see the graphs! (that was about the extent of my test plan, if there's any other things to poke at I can)

### CODE REVIEW

File: topmodel/file_system.py [hunk 1]
```diff
@@ -55,7 +57,31 @@ def write_file(self, path, data):
         key = self.bucket.get_key(self.subdirectory + path)
         if key is None:
             key = self.bucket.new_key(self.subdirectory + path)
-        key.set_contents_from_string(data)
+
+        # Take the data (a giant string of JSON), and turn it into a file
+        # stream we can read from
+        f = StringIO.StringIO()
+        f.write(data)
+        f.seek(0)
```
Comment (mduvall) [inline topmodel/file_system.py]: So the `write` above puts the pointer at EOF, this puts it back at the beginning. (`initiate_multipart_upload` doesn't handle that very well)

File: topmodel/file_system.py [hunk 2]
```diff
@@ -55,7 +57,31 @@ def write_file(self, path, data):
...
+        chunks = int(math.ceil(size/float(chunk_size)))
+
+        # Start the multipart upload
+        upload = self.bucket.initiate_multipart_upload(key)
+
+        try:
+            # Upload chunk by chunk, using the StringIO object instantiated above
+            # we can reference the offset we want to upload.
+            for i in range(chunks):
+                upload.upload_part_from_file(f, part_num=i+1)
```
Comment (mduvall) [inline topmodel/file_system.py]: `part_num` starts from 1, chunks.

File: topmodel/file_system.py [hunk 3]
```diff
@@ -55,7 +57,31 @@ def write_file(self, path, data):
...
+        except:
+            upload.cancel_upload()
```
Comment (mduvall) [inline topmodel/file_system.py]: We have to close the connection manually if `upload_part_from_file` throws an exception :turtle:

**Comment by mduvall** [comment]
cc @jocelynsarah (more `topmodel` :eyes:)

File: topmodel/file_system.py [hunk 4]
```diff
@@ -55,7 +57,31 @@ def write_file(self, path, data):
         key = self.bucket.get_key(self.subdirectory + path)
         if key is None:
             key = self.bucket.new_key(self.subdirectory + path)
-        key.set_contents_from_string(data)
+
+        # Take the data (a giant string of JSON), and turn it into a file
```
Comment (mduvall) [inline topmodel/file_system.py]: Err I'll fix this comment, it's not all giant strings of JSON.

Comment (alyssafrazee) [inline topmodel/file_system.py]: :white_check_mark:

File: topmodel/file_system.py [hunk 5]
```diff
@@ -55,7 +57,32 @@ def write_file(self, path, data):
...
+        # stream we can read from
+        f = StringIO.StringIO()
+        f.write(data)
+        f.seek(0)
+
+        # Count the number of chunks we're going to have to upload
+        size = len(data)
+        chunk_size = 1024 * 1024 * 50
```
Comment (alyssafrazee) [inline topmodel/file_system.py]: Can you add a comments saying how large this is? ("default 50 Mb" or whatever?)

**Comment by alyssafrazee** [comment]
This looks great. Does the uploaded file have the same format as it did before? (I assume it does, since your tests worked etc, but it's not totally clear to me from the [`complete_upload` docs](http://boto.cloudhackers.com/en/latest/ref/s3.html) how the upload is "completed"). 

:ship:

File: topmodel/file_system.py (hunk 5 above)
Comment (mduvall) [inline topmodel/file_system.py]: Ah for sure!

**Pull request merged**